### PR TITLE
e2e: parallel build of current and WIP version bundle versions

### PR DESCRIPTION
**Pull request by fgimenez**
With these changes the version bundle version is obtained from the version bundles defined in the `service` package. In this case this approach makes more sense than getting the yaml output form `aws-operator version` and parsing it, using the method in the library there's no need to write a file, read it and unmarshal the yaml.

I've added modifications to the circleCI config file to run builds in parallel for the current and WIP version bundle versions, we can extend this to test other deprecated versions if needed. If the WIP version is not found the fact is logged and the build succeeds.

### CODE REVIEW

**Comment by xh3b4sd** [comment]
FYI the tests fail.

File: integration/operator_test.go [hunk 1]
```diff
@@ -98,6 +99,16 @@ func TestMain(m *testing.M) {
...
 
+	if os.Getenv("VERSION_BUNDLE_VERSION") == "" {
+		version := f.GetVersionBundleVersion(service.NewVersionBundles(), os.Getenv("TESTED_VERSION"))
+		if version == "" {
+			log.Printf("No version bundle version for TESTED_VERSION %q", os.Getenv("TESTED_VERSION"))
+			os.Exit(0)
+		}
+		os.Setenv("VERSION_BUNDLE_VERSION", version)
+	}
+	log.Printf("VERSION_BUNDLE_VERSION %v", os.Getenv("VERSION_BUNDLE_VERSION"))
```
Comment (kopiczko) [inline integration/operator_test.go]: I think it would be nice to encapsulate whole that thing in GetVersionBundleVersion.

**Comment by xh3b4sd** [comment]
IMO we have to test all the versions we support all the time, even when they are deprecated. One could argue that they have been tested already and that they are frozen anyway, but anything at runtime can just fail. I think about a CoreOS or Kubernetes upgrade for instance. Then we have to make sure that even the very last guest cluster running at the last deprecated version is still functional and fine. I think we can reduce risk and improve stability when just testing all the versions we have.

File: integration/framework.go [hunk 2]
```diff
@@ -457,3 +458,16 @@ func (f *framework) ApplyAWSConfigPatch(patch []PatchSpec, clusterName string) e
...
+
+func (f *framework) GetVersionBundleVersion(bundle []versionbundle.Bundle, vType string) string {
+	var output string
+	for _, v := range bundle {
+		if (vType == "current" || vType == "") && !v.Deprecated && !v.WIP {
+			output = v.Version
+		}
+		if vType == "wip" && v.WIP {
+			output = v.Version
+		}
```
Comment (kopiczko) [inline integration/framework.go]: What if there are multiple WIP or non-deprecated non-wip versions? It would be nice to have some guarantees.

**Comment by fgimenez** [comment]
@xh3b4sd exactly, that's the goal. This PR introduces the changes to make it possible, let's see how it works with two versions and when we make sure that all is good extend to cover all of them.

File: integration/framework.go (hunk 2 above)
Comment (xh3b4sd) [reply integration/framework.go]: This looks definitely not that generic. What about having a circle template and just render it according to all version bundles available? I do not see a generic future direction with these conditionals.

Comment (fgimenez) [reply integration/framework.go]: > What about having a circle template and just render it according to all version bundles available?

I'm afraid that wouldn't work, currently we need a static circleCI config file committed to the repo for the github integration to work. Looks like there's already a feature request for dynamic configuration https://discuss.circleci.com/t/dynamic-2-0-configuration/19287 once this is in place we can get rid of these hardcoded checks. Until then I was planning to update the e2e circleci config according to the existent version bundles but after we make sure that two concurrent builds work as expected.

Comment (fgimenez) [reply integration/framework.go]: > Until then I was planning to update the e2e circleci config according to the existent version bundles but after we make sure that two concurrent builds work as expected.

For this we would just need to have enough jobs in the config (no need to set the actual version bundle versions in the CI config, but making sure that there enough for the current bundle), the dynamic retrieval would pick up the version bundle version for the job index.

**Comment by fgimenez** [comment]
Regarding the parallel execution, the IPAM wasn't working, the source of truth is EC2 api, by the time the builds start they all query the API and get the next free CIDR, all of them the same. It takes time for the stack to be up only when the VPC resource has been created, the next free CIDR would be right.

Instead of that now a different VPC and route tables are created for each build, and the guest VPC and its subnets get fixed CIDRs. This way we prevent overlaps and can run as many parallel builds as needed. These host resources are created in a separate stack and are removed after the build is done.

File: integration/operator_test.go [hunk 3]
```diff
@@ -98,8 +102,24 @@ func TestMain(m *testing.M) {
...
+	}
+	log.Printf("TESTED_VERSION %v", os.Getenv("TESTED_VERSION"))
+	log.Printf("VERSION_BUNDLE_VERSION %v", os.Getenv("VERSION_BUNDLE_VERSION"))
+
 	c = newAWSClient()
 
+	if err := createHostPeerVPC(); err != nil {
+		log.Printf("unexpected error: %v\n", err)
+		os.Exit(1)
+	}
```
Comment (kopiczko) [inline integration/operator_test.go]: Is this change related? Can this be split into another PR?

Comment (fgimenez) [reply integration/operator_test.go]: Yes, it is required for the parallel execution, the PR involves getting WIP and current versions and running them, do you think it is better to split?

File: integration/operator_test.go (hunk 1 above)
Comment (fgimenez) [reply integration/operator_test.go]: mm not sure, all the code outside the `_test` file is going soon to the `e2e-harness` project so that other projects can use it as a library. I'd prefer to not move code there with side effects, like setting an env var like in this case, even the env var name can depend on the project, WDYT?

File: integration/operator_test.go (hunk 3 above)
Comment (kopiczko) [reply integration/operator_test.go]: I would appreciate that.

Comment (fgimenez) [reply integration/operator_test.go]: ok done in #729, once that one lands I'll merge master here, the changes there are required here.

Comment (kopiczko) [reply integration/operator_test.go]: 👍 I reviewed that one already. Ping me when this one is ready to review.

Comment (fgimenez) [reply integration/operator_test.go]: @kopiczko ready for review, now only changes related to parallel execution here.

File: integration/operator_test.go [hunk 4]
```diff
@@ -334,3 +328,83 @@ func removeWorker(clusterName string) error {
...
+func createHostPeerVPC() error {
+	log.Printf("Creating Host Peer VPC stack")
+
+	hostVPCStack := `AWSTemplateFormatVersion: 2010-09-09
+Description: CI Host Stack with Peering VPC and route tables
+Resources:
+  VPC:
+    Type: AWS::EC2::VPC
+    Properties:
+      CidrBlock: 10.11.0.0/16
```
Comment (kopiczko) [inline integration/operator_test.go]: For my understanding: is it reserved for testing? Or CF is smart enough to not create overlapping route tables?

Comment (fgimenez) [reply integration/operator_test.go]: This is the CIDR used for testing. All the test peer accepter VPCs will have the same, they are isolated by definition. 

This prevents collisions in the routes used for peering, if we use the same VPC for all the builds (as we did before and as we do in production), then the routes used for each of the guest clusters  must have a different destination, we can't have two routes with the same CIDR destination in the same route table. We can solve this by assigning each guest a different, non-overlapping CIDR, but this didn't work for parallel builds because they all request a free CIDR at the same time, and they all get the same.

By using a new, separate peer accepter VPC for each build (and separate route tables associated to it just created for each build) we can make sure that the routes to the guest subnets have unique target CIDR, because both the accepter VPC and the route tables only live during the build. We can even assign all the guest VPCs the same CIDR, no need for IPAM at all.
